src/main/java/data/Recipe.java

```
@@ -0,0 +1,115 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class Recipe {
+
+  protected String name;
+  protected String description;
+  protected List<Step> steps;
+  protected List<SpinOff> spinOffs;
+
+  public Recipe(Recipe original) {
+    this.name = original.name;
+    this.description = original.description;
+    this.steps = original.steps;
+    this.spinOffs = original.spinOffs;
+  }
+  
+  public Recipe(String name, String description, List<Step> steps) {
+    this.name = name;
+    this.description = description;
+    this.steps = steps;
+    this.spinOffs = new LinkedList<>();
+  }
+
+  public void appendStep(Step newStep) {
+    steps.add(newStep);
+  }
+
+  public void addStep(int position, Step newStep) {
+    if (position < 0) {
+      System.err.println("Position " + position + " out of bounds, failed to add step [" + newStep + "]");
+    } else if (position >= steps.size()) {
+      System.err.println("Position " + position + " out of bounds, appending step [" + newStep + "]");
```

YOu should not log to error and add a new step. You could also throw an error or return some value along with logging to indicate that the addition of new step was not successful. This can then be used to propagate to user in an appropriate way.